test/bot_tests/bots_test.py failed

Details [here](https://github.com/ALLAN-DIP/baseline_bots/actions/runs/3395300259/jobs/5645002468)
Error:
```
=================================== FAILURES ===================================
________________________________ TestRPBot.test ________________________________

self = <bots_test.TestRPBot object at 0x7f3c0d7bf320>

    def test(self):
        game_play = GamePlay(None, [RandomProposerBot_AsyncBot, RandomProposerBot_AsyncBot], 3, True)
>       actions, done = game_play.step()

tests/bot_tests/bots_test.py:18: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/data/env3.7/lib/python3.7/site-packages/tornado/gen.py:776: in run
    yielded = self.gen.throw(*exc_info)  # type: ignore
tests/bot_tests/gameplay_framework.py:90: in step
    ret_dict = yield bot(rcvd_messages)
/data/env3.7/lib/python3.7/site-packages/tornado/gen.py:769: in run
    value = future.result()
/data/env3.7/lib/python3.7/site-packages/tornado/gen.py:782: in run
    yielded = self.gen.send(value)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <baseline_bots.bots.random_proposer_bot.RandomProposerBot_AsyncBot object at 0x7f3c0cbf3470>
rcvd_messages = []

    @gen.coroutine
    def __call__(self, rcvd_messages):
        """
        :return: dict containing messages and orders
        """
        messages = yield self.gen_messages(rcvd_messages)
    
        if messages and messages.messages:
            for msg in messages.messages:
                msg_obj = Message(
                    sender=self.power_name,
                    recipient=msg['recipient'],
                    message=msg['message'],
>                   phase=game.get_current_phase(),
                )
E               NameError: name 'game' is not defined

src/baseline_bots/bots/random_proposer_bot.py:104: NameError
=========================== short test summary info ============================
FAILED tests/bot_tests/bots_test.py::TestRPBot::test - NameError: name 'game'...
============================== 1 failed in 0.57s ===============================
The command '/bin/sh -c pytest tests/bot_tests/bots_test.py' returned a non-zero code: 1
Error: Process completed with exit code 1.
```

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -1,7 +1,7 @@
 name: CI test
 on:
   pull_request:
-    branches: [ main, test_CI ]
+    branches: [ main ]
   push:
     branches: [ main, test_CI ]
 
@@ -19,6 +19,9 @@ jobs:
 
     - name: "Build Docker Container"
         run: |
+          wget https://f002.backblazeb2.com/file/ppaquette-public/benchmarks/neurips2019-sl_model.zip
+          mkdir bot_neurips2019-sl_model
+          unzip neurips2019-sl_model.zip -d bot_neurips2019-sl_model/
           docker build --no-cache -t allan_test_env .
 
     - name: "Run tests"

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -1,7 +1,7 @@
 name: CI test
 on:
   pull_request:
-    branches: [ main ]
+    branches: [ main, test_CI ]
   push:
     branches: [ main, test_CI ]
 

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -27,4 +27,4 @@ jobs:
 
       - name: "Run Tests"
         run: |
-          docker run --rm ci_image /bin/bash -c "/model/src/model_server/baseline_bots/containers/allan_dip_bot/run_model_server.sh & pytest --ignore=tests/bot_tests/SOA_test.py --ignore=tests/bot_tests/bots_test.py --ignore=tests/bot_tests/dipnet.py"
+          docker run --rm ci_image /bin/bash -c "/model/src/model_server/baseline_bots/containers/allan_dip_bot/run_model_server.sh & pytest --ignore=tests/bot_tests/SOA_test.py"

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -20,11 +20,9 @@ jobs:
       
       - name: "Build Docker Container"
         run: |
-          wget https://f002.backblazeb2.com/file/ppaquette-public/benchmarks/neurips2019-sl_model.zip
-          mkdir bot_neurips2019-sl_model
-          unzip neurips2019-sl_model.zip -d bot_neurips2019-sl_model/
           docker build --no-cache -t ci_image .
 
       - name: "Run Tests"
         run: |
-          docker run --rm ci_image /bin/bash -c "/model/src/model_server/baseline_bots/containers/allan_dip_bot/run_model_server.sh & pytest --ignore=tests/bot_tests/SOA_test.py"
+          docker run --rm ci_image /bin/bash -c "/model/src/model_server/baseline_bots/containers/allan_dip_bot/run_model_server.sh & \
+          pytest --ignore=tests/bot_tests/SOA_test.py"

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -24,7 +24,4 @@ jobs:
 
     - name: "Run tests"
       run: |
-        pytest tests/utils_test.py
-        pytest tests/bot_tests/bots_test.py
-        # pytest tests/randomize_order_test.py
-        # pytest tests/bot_tests/soa_bot_functions_test.py
+        docker run --rm allan_test_env /bin/bash -c "pytest --ignore=tests/bot_tests/SOA_test.py --ignore=tests/randomize_order_test.py --ignore=tests/bot_tests/dipnet.py"

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -17,10 +17,9 @@ jobs:
       with:
         python-version: '3.7'
 
-    - name: "Build"
-      run: |
-        pip install -r requirements.txt
-        pip install -e .
+    - name: "Build Docker Container"
+        run: |
+          docker build --no-cache -t allan_test_env .
 
     - name: "Run tests"
       run: |

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -19,7 +19,6 @@ jobs:
 
     - name: "Build Docker Container"
         run: |
-          echo "Building Docker Container"
           docker build --no-cache -t allan_test_env .
 
     - name: "Run tests"

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -28,5 +28,3 @@ jobs:
       - name: "Run Tests"
         run: |
           docker run --rm ci_image /bin/bash -c "/model/src/model_server/baseline_bots/containers/allan_dip_bot/run_model_server.sh & pytest --ignore=tests/bot_tests/SOA_test.py --ignore=tests/bot_tests/bots_test.py --ignore=tests/bot_tests/dipnet.py"
-
-docker run --rm ci_image /bin/bash -c "/model/src/model_server/baseline_bots/containers/allan_dip_bot/run_model_server.sh & python t.py"

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -19,6 +19,7 @@ jobs:
 
     - name: "Build Docker Container"
         run: |
+          echo "Building Docker Container"
           docker build --no-cache -t allan_test_env .
 
     - name: "Run tests"

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,42 +1,62 @@
-FROM ubuntu:18.04
-RUN mkdir research
-RUN mkdir baseline_bots
-RUN rm /bin/sh && ln -s /bin/bash /bin/sh
-RUN apt-get update
-RUN apt-get -y upgrade
-RUN apt-get -y install python3.7
-RUN touch ~/.bash_profile && echo "alias python='/usr/bin/python3.7'" > ~/.bash_profile && source ~/.bash_profile
-RUN apt -y install software-properties-common
-RUN add-apt-repository ppa:deadsnakes/ppa -y
+FROM pcpaquette/tensorflow-serving:20190226
+
+WORKDIR /model/src/model_server
+
+RUN apt-get -y update
 RUN apt-get -y install git
-RUN apt-get -y install python3-pip
-RUN apt-get install wget
-RUN apt-get install -y build-essential libssl-dev uuid-dev libgpgme11-dev libseccomp-dev pkg-config squashfs-tools
-RUN pip3 install --upgrade setuptools
-RUN git clone https://github.com/SHADE-AI/research.git && cd research && pip3 install -r requirements.txt
-COPY . /research/ 
-RUN cd $HOME
-RUN git clone https://github.com/ALLAN-DIP/baseline_bots.git && cd baseline_bots
-ADD . baseline_bots/
-RUN cd $HOME
+RUN apt-get -y install vim
+RUN apt-get -y install curl
+RUN apt-get -y install htop
+RUN apt-get -y install lsof
+
+# Copy SL model
+RUN mkdir /model/src/model_server/bot_neurips2019-sl_model
+COPY bot_neurips2019-sl_model /model/src/model_server/bot_neurips2019-sl_model 
+COPY run_model_server.sh /model/src/model_server/run_model_server.sh
+RUN chmod 777 /model/src/model_server/run_model_server.sh
+RUN chmod -R 777 /model/src/model_server/bot_neurips2019-sl_model
+
+# TODO: Get this to work for RL model as well
+
+# Clone repos
+RUN git clone https://github.com/SHADE-AI/diplomacy.git
+RUN git clone https://github.com/SHADE-AI/research.git
+RUN git clone https://github.com/ALLAN-DIP/baseline_bots.git
+
+# Environment variables
 ENV PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=cpp
 ENV PYTHONIOENCODING=utf-8
 ENV LANG=en_CA.UTF-8
 ENV PYTHONUNBUFFERED=1
 ENV PATH=/data/env3.7/bin:$PATH
 
-ENV VERSION=v3.2.0
-ENV GO_VERSION=1.12.5 OS=linux ARCH=amd64
-ENV GOPATH=$HOME/.go
-ENV PATH=/usr/local/go/bin:${PATH}:${GOPATH}/bin
-RUN wget -nv https://dl.google.com/go/go$GO_VERSION.$OS-$ARCH.tar.gz && tar -C /usr/local -xzf go$GO_VERSION.$OS-$ARCH.tar.gz && rm -f go$GO_VERSION.$OS-$ARCH.tar.gz && mkdir -p $GOPATH && go get github.com/golang/dep/cmd/dep && mkdir -p $GOPATH/src/github.com/sylabs && cd $GOPATH/src/github.com/sylabs && git clone https://github.com/sylabs/singularity.git && cd singularity && git checkout v3.2.0 &&./mconfig -p /usr/local &&cd ./builddir && make && make install
+# Default batching parameters, can override with docker run -e 
+ENV MAX_BATCH_SIZE=128
+ENV BATCH_TIMEOUT_MICROS=250000
+ENV MAX_ENQUEUED_BATCHES=1024
+ENV NUM_BATCH_THREADS=8
+ENV PAD_VARIABLE_LENGTH_INPUTS='true'
 
-RUN cd $HOME/baseline_bots
-RUN pip3 install git+https://github.com/trigaten/DAIDE
+# Avoid git issues
+RUN git config --global --add safe.directory /model/src/model_server/diplomacy
+RUN git config --global --add safe.directory /model/src/model_server/research
+RUN git config --global --add safe.directory /model/src/model_server/baseline_bots
 
-RUN chmod 777 dip_ui_bot_launcher.py
-ENV PATH=/baseline_bots/:$PATH
-ENV PYTHONPATH=$PYTHONPATH:$ROOT/research/
+# Avoid pip issues
+RUN pip install --upgrade pip
 
+# Install diplomacy research requirements
+WORKDIR /model/src/model_server/research
+RUN sed -i 's/gym>/gym=/g'  requirements.txt
+RUN pip install -r requirements.txt
 
-ENTRYPOINT ["python", "dip_ui_bot_launcher.py","-H","hostname","-p","powers","-B", "bots", "-g", "gameid"]
+# Install baseline_bots requirements
+WORKDIR /model/src/model_server/baseline_bots
+RUN pip install -r requirements.txt
+RUN pip install -e .
+
+# Script executors
+COPY run_bot.py /model/src/model_server/baseline_bots/run_bot.py
+COPY run.sh /model/src/model_server/baseline_bots/run.sh
+RUN chmod 777 /model/src/model_server/baseline_bots/run_bot.py
+RUN chmod 777 /model/src/model_server/baseline_bots/run.sh

--- a/Dockerfile
+++ b/Dockerfile
@@ -12,8 +12,6 @@ RUN apt-get -y install lsof
 # Copy SL model
 RUN mkdir /model/src/model_server/bot_neurips2019-sl_model
 COPY bot_neurips2019-sl_model /model/src/model_server/bot_neurips2019-sl_model 
-COPY run_model_server.sh /model/src/model_server/run_model_server.sh
-RUN chmod 777 /model/src/model_server/run_model_server.sh
 RUN chmod -R 777 /model/src/model_server/bot_neurips2019-sl_model
 
 # TODO: Get this to work for RL model as well
@@ -54,9 +52,3 @@ RUN pip install -r requirements.txt
 WORKDIR /model/src/model_server/baseline_bots
 RUN pip install -r requirements.txt
 RUN pip install -e .
-
-# Script executors
-COPY run_bot.py /model/src/model_server/baseline_bots/run_bot.py
-COPY run.sh /model/src/model_server/baseline_bots/run.sh
-RUN chmod 777 /model/src/model_server/baseline_bots/run_bot.py
-RUN chmod 777 /model/src/model_server/baseline_bots/run.sh

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,7 +1,6 @@
 FROM pcpaquette/tensorflow-serving:20190226
 
-WORKDIR /model/src/model_server
-
+# install needed packages
 RUN apt-get -y update
 RUN apt-get -y install git
 RUN apt-get -y install vim
@@ -9,20 +8,20 @@ RUN apt-get -y install curl
 RUN apt-get -y install htop
 RUN apt-get -y install lsof
 
-# Copy SL model
-RUN mkdir /model/src/model_server/bot_neurips2019-sl_model
-COPY bot_neurips2019-sl_model /model/src/model_server/bot_neurips2019-sl_model 
-RUN chmod -R 777 /model/src/model_server/bot_neurips2019-sl_model
+WORKDIR /model/src/model_server
 
-# TODO: Get this to work for RL model as well
+# Copy SL model
+RUN wget https://f002.backblazeb2.com/file/ppaquette-public/benchmarks/neurips2019-sl_model.zip
+RUN mkdir /model/src/model_server/bot_neurips2019-sl_model
+RUN unzip neurips2019-sl_model.zip -d /model/src/model_server/bot_neurips2019-sl_model 
+RUN chmod -R 777 /model/src/model_server/bot_neurips2019-sl_model
 
 # Clone repos
 RUN git clone https://github.com/SHADE-AI/diplomacy.git
 RUN git clone https://github.com/SHADE-AI/research.git
 
+# copy baseline bots code into the docker image
 RUN mkdir /model/src/model_server/baseline_bots
-
-
 COPY . /model/src/model_server/baseline_bots
 
 # Environment variables
@@ -57,5 +56,7 @@ WORKDIR /model/src/model_server/baseline_bots
 RUN pip install -r requirements.txt
 RUN pip install -e .
 
+# allow the tf server to be run
 RUN chmod 777 /model/src/model_server/baseline_bots/containers/allan_dip_bot/run_model_server.sh
+# add diplomacy research to python path
 ENV PYTHONPATH=/model/src/model_server/research:$PYTHONPATH

--- a/tests/bot_tests/SOA_test.py
+++ b/tests/bot_tests/SOA_test.py
@@ -1,9 +1,9 @@
 """unit tests for smart order accepter bot"""
+import tornado
 from diplomacy import Game, Message
 from diplomacy_research.utils.cluster import start_io_loop, stop_io_loop
 from gameplay_framework_async import GamePlayAsync
 from tornado import gen
-import tornado
 
 from baseline_bots.bots.baseline_bot import BaselineBot, BaselineMsgRoundBot
 from baseline_bots.bots.random_proposer_bot import RandomProposerBot_AsyncBot
@@ -466,7 +466,7 @@ if __name__ == "__main__":
 
 @gen.coroutine
 def main():
-    """ Plays a local game with 7 bots """
+    """Plays a local game with 7 bots"""
     game = Game()
     player = SmartOrderAccepterBot("FRANCE", game)
     f = open("demofile2.txt", "a")

--- a/tests/bot_tests/SOA_test.py
+++ b/tests/bot_tests/SOA_test.py
@@ -3,6 +3,7 @@ from diplomacy import Game, Message
 from diplomacy_research.utils.cluster import start_io_loop, stop_io_loop
 from gameplay_framework_async import GamePlayAsync
 from tornado import gen
+import tornado
 
 from baseline_bots.bots.baseline_bot import BaselineBot, BaselineMsgRoundBot
 from baseline_bots.bots.random_proposer_bot import RandomProposerBot_AsyncBot
@@ -88,10 +89,11 @@ class TestSOABot:
 
         stop_io_loop()
 
-    @gen.coroutine
+    @tornado.testing.gen_test
     def test_play(self):
         game = Game()
         soa_bot1 = SmartOrderAccepterBot("FRANCE", game)
+
         soa_bot2 = SmartOrderAccepterBot("RUSSIA", game)
         game_play = GamePlayAsync(
             game,
@@ -460,3 +462,14 @@ class TestSOABot:
 if __name__ == "__main__":
     soa_test = TestSOABot()
     soa_test.test()
+
+
+@gen.coroutine
+def main():
+    """ Plays a local game with 7 bots """
+    game = Game()
+    player = SmartOrderAccepterBot("FRANCE", game)
+    f = open("demofile2.txt", "a")
+    f.write("Now the file has more content!")
+    f.close()
+    stop_io_loop()

--- a/tests/bot_tests/SOA_test.py
+++ b/tests/bot_tests/SOA_test.py
@@ -3,9 +3,7 @@ import tornado
 from diplomacy import Game, Message
 from diplomacy_research.utils.cluster import start_io_loop, stop_io_loop
 from gameplay_framework_async import GamePlayAsync
-from tornado import gen
-import tornado
-from tornado import testing
+from tornado import gen, testing
 from tornado.testing import AsyncTestCase
 
 from baseline_bots.bots.baseline_bot import BaselineBot, BaselineMsgRoundBot
@@ -121,7 +119,6 @@ class TestSOABot(AsyncTestCase):
 
         stop_io_loop()
 
-    
     @gen.coroutine
     def score_stance(self):
         # score-based

--- a/tests/bot_tests/simple_SOA_test.py
+++ b/tests/bot_tests/simple_SOA_test.py
@@ -1,9 +1,9 @@
 """unit tests for smart order accepter bot"""
 from diplomacy import Game, Message
 from diplomacy_research.utils.cluster import start_io_loop, stop_io_loop
-from tornado import gen
-from tornado import testing
+from tornado import gen, testing
 from tornado.testing import AsyncTestCase
+
 from baseline_bots.bots.baseline_bot import BaselineBot, BaselineMsgRoundBot
 from baseline_bots.bots.random_proposer_bot import RandomProposerBot_AsyncBot
 from baseline_bots.bots.smart_order_accepter_bot import SmartOrderAccepterBot

--- a/tests/bot_tests/simple_SOA_test.py
+++ b/tests/bot_tests/simple_SOA_test.py
@@ -1,0 +1,34 @@
+"""unit tests for smart order accepter bot"""
+from diplomacy import Game, Message
+from diplomacy_research.utils.cluster import start_io_loop, stop_io_loop
+from tornado import gen
+from tornado import testing
+from tornado.testing import AsyncTestCase
+from baseline_bots.bots.baseline_bot import BaselineBot, BaselineMsgRoundBot
+from baseline_bots.bots.random_proposer_bot import RandomProposerBot_AsyncBot
+from baseline_bots.bots.smart_order_accepter_bot import SmartOrderAccepterBot
+from baseline_bots.parsing_utils import (
+    daide_to_dipnet_parsing,
+    dipnet_to_daide_parsing,
+    parse_proposal_messages,
+)
+from baseline_bots.utils import (
+    MessagesData,
+    OrdersData,
+    get_best_orders,
+    get_order_tokens,
+    get_other_powers,
+    get_state_value,
+    parse_alliance_proposal,
+    parse_arrangement,
+    parse_PRP,
+)
+
+
+class TestSOABot(AsyncTestCase):
+    @testing.gen_test
+    def test_play(self):
+        game = Game()
+        soa_bot = SmartOrderAccepterBot("FRANCE", game)
+        messages = MessagesData().add_message("FRANCE", "A PAR - BUR")
+        soa_bot.send_message("FRANCE", messages)

